File: app/src/main/java/com/nao4j/subtrol/security/VerifyTokenFilter.java
package com.nao4j.subtrol.security;

import com.nao4j.subtrol.dto.UserCredentials;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.GenericFilterBean;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Set;

import static io.jsonwebtoken.security.Keys.hmacShaKeyFor;
import static java.time.Instant.now;
import static java.util.Collections.emptySet;
import static javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED;

@RequiredArgsConstructor
public class VerifyTokenFilter extends GenericFilterBean {

    private static final String AUTH_HEADER_NAME = "Authorization";
    private static final String TOKEN_PREFIX = "Bearer";

    private final String secret;

    @Override
    public void doFilter(final ServletRequest req, final ServletResponse res, final FilterChain filterChain)
            throws IOException, ServletException {
        final var request = (HttpServletRequest) req;
        final var response = (HttpServletResponse) res;
        try {
            final var authentication = verifyToken(request);
            if (authentication != null) {
                SecurityContextHolder.getContext().setAuthentication(authentication);
            } else {
                SecurityContextHolder.getContext().setAuthentication(null);
            }
            filterChain.doFilter(req, res);
        } catch (final JwtException e) {
            response.setStatus(SC_UNAUTHORIZED);
        } finally {
            SecurityContextHolder.getContext().setAuthentication(null);
        }
    }

    private Authentication verifyToken(final HttpServletRequest request) {
        final var token = request.getHeader(AUTH_HEADER_NAME);
        if ((token != null) && !token.isEmpty()) {
            final var user = parseUserFromToken(token.replace(TOKEN_PREFIX, "").trim());
            if (user != null) {
                return new UserAuthentication(user);
            }
        }
        return null;
    }

    private UserCredentials parseUserFromToken(final String token) {
        final var claims = Jwts.parser()
                .setSigningKey(hmacShaKeyFor(secret.getBytes()))   // todo: bad security solution
                .parseClaimsJws(token)
                .getBody();
        if (now().isAfter(claims.getExpiration().toInstant())) {
            return null;
        }
        final var userId = (String) claims.get("userId");
        final Set<String> roles;
        if (claims.get("roles") != null) {
            roles = Set.of(((String) claims.get("roles")).split(","));
        } else {
            roles = emptySet();
        }
        if ((userId != null) && !roles.isEmpty()) {
            return new UserCredentials(userId, "", roles);
        } else {
            return null;
        }
    }

}

File: src/main/kotlin/com/nao4j/subtrol/security/VerifyTokenFilter.kt
package com.nao4j.subtrol.security

import com.nao4j.subtrol.dto.UserCredentials
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.filter.GenericFilterBean
import java.io.IOException
import java.time.Instant
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED

private const val AUTH_HEADER_NAME = "Authorization"
private const val TOKEN_PREFIX = "Bearer"

class VerifyTokenFilter(private val secret: String) : GenericFilterBean() {

    @Throws(IOException::class, ServletException::class)
    override fun doFilter(req: ServletRequest, res: ServletResponse, filterChain: FilterChain) {
        val request = req as HttpServletRequest
        val response = res as HttpServletResponse
        try {
            val authentication = verifyToken(request)
            if (authentication != null) {
                SecurityContextHolder.getContext().authentication = authentication
            } else {
                SecurityContextHolder.getContext().authentication = null
            }
            filterChain.doFilter(req, res)
        } catch (e: JwtException) {
            response.status = SC_UNAUTHORIZED
        } finally {
            SecurityContextHolder.getContext().authentication = null
        }
    }

    private fun verifyToken(request: HttpServletRequest): Authentication? {
        val token = request.getHeader(AUTH_HEADER_NAME)
        if (token != null && !token.isEmpty()) {
            val user = parseUserFromToken(token.replace(TOKEN_PREFIX, "").trim { it <= ' ' })
            if (user != null) {
                return UserAuthentication(user)
            }
        }
        return null
    }

    private fun parseUserFromToken(token: String): UserCredentials? {
        val claims = Jwts.parser().setSigningKey(secret).parseClaimsJws(token).body
        if (Instant.now() > claims.expiration.toInstant()) {
            return null
        }
        val userId = claims["userId"] as String?
        val username = claims.subject
        val roles: Set<String> = if (claims["roles"] != null) {
            (claims["roles"] as String).split(",").toSet()
        } else {
            emptySet()
        }
        return if (userId != null && !roles.isEmpty()) {
            UserCredentials(userId, username, "", roles)
        } else {
            null
        }
    }

}
